include/zio/zio_buffer.h

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2019 Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZIO_BUFFER_H_
+#define ZEPHYR_INCLUDE_ZIO_BUFFER_H_
+
+#include <zephyr/types.h>
+#include <kernel.h>
+
+struct zio_buffer {
+  struct k_poll_signal signal;
+  struct device *device;
+  bool timestamps;
+  bool circular;
+  struct zio_fifo *fifo;
+  u32_t dropped_samples;
```

Comment:
I was thinking about how to track missed samples. It may not be obvious (or even possible) to detect with many types of sensors, but what we can do to expose some sort of metric to check at the application level is to have a **sample counter**, where every time the driver updates the channel content, the counter is incremented by 1. By comparing the channel counter value of the current sample read from the driver, versus the value from the last time the application read data, it's easy to determine not just if samples were missed, but the number of samples missed, and something like a 16 bit value that wraps around should be enough for this even with relatively high sample rates.

Reply:
I think this makes sense. When a hardware fifo is involved we might get an overflow interrupt signal and status but thats about it, we probably won't know how many samples we've actually lost

Reply:
Thinking more about this there's a few possible reason for missed/dropped samples...

* A circular software fifo was chosen and overwrites of old values occured.
* A hardware fifo is provided and overflows notified about, depending on the hardware we may/may not know how many samples are lost

Given its implementation specific we could have both an optional flag that is set when samples have been lost since the last zio_buf was triggered and an optional count.

The builtin software fifo backed zio_buf would be able to deal with both. Hardware backed fifos might optionally implement zero or more of those features.

So to deal with that we should probably add a feature flags attribute to zio_buf as a flag notifying the zio_buf user there were samples lost if that feature is available, and the count of samples lost if that feature is available.

Reply:
A flag is probably a good idea in general anyway and not overly expensive. Good points though.